### PR TITLE
Fixed bug in remove_duplicated_spikes with 'random' method

### PR DESCRIPTION
**Pull request by DradeAW**
With `spike_train = np.array([250, 850, 851, 852, 950])` and `method="random"`, sometimes the indices of duplicated spikes found was `[1, 2, 3]`, which it shouldn't be.

This should fix it.

### CODE REVIEW

File: spikeinterface/curation/curation_tools.py [hunk 1]
```diff
@@ -33,6 +32,21 @@ def _find_duplicated_spikes_numpy(spike_train: np.ndarray, censored_period: int,
...
 
+def _find_duplicated_spikes_random(spike_train: np.ndarray, censored_period: int, seed: int) -> np.ndarray:
+    rand_state = np.random.get_state()
+    np.random.seed(seed)
+
+    indices_of_duplicates = []
+    while not np.all(np.diff(np.delete(spike_train, indices_of_duplicates)) > censored_period):
+        duplicates, = np.where(np.diff(spike_train) <= censored_period)
+        duplicates = np.unique(np.concatenate((duplicates, duplicates + 1)))
+        duplicate = np.random.choice(duplicates)
```
Comment (samuelgarcia) [inline spikeinterface/curation/curation_tools.py:43]: `np.random.RandomState(seed=seed).choice()` this avoid to make variable for random state.

File: spikeinterface/curation/curation_tools.py [hunk 2]
```diff
@@ -33,6 +32,21 @@ def _find_duplicated_spikes_numpy(spike_train: np.ndarray, censored_period: int,
     return indices_of_duplicates
 
 
+def _find_duplicated_spikes_random(spike_train: np.ndarray, censored_period: int, seed: int) -> np.ndarray:
+    rand_state = np.random.get_state()
+    np.random.seed(seed)
```
Comment (samuelgarcia) [inline spikeinterface/curation/curation_tools.py:37]: You can remove this with the bellow approahc

File: spikeinterface/curation/curation_tools.py [hunk 3]
```diff
@@ -33,6 +32,21 @@ def _find_duplicated_spikes_numpy(spike_train: np.ndarray, censored_period: int,
...
+
+    indices_of_duplicates = []
+    while not np.all(np.diff(np.delete(spike_train, indices_of_duplicates)) > censored_period):
+        duplicates, = np.where(np.diff(spike_train) <= censored_period)
+        duplicates = np.unique(np.concatenate((duplicates, duplicates + 1)))
+        duplicate = np.random.choice(duplicates)
+        indices_of_duplicates.append(duplicate)
+
+    np.random.set_state(rand_state)
```
Comment (samuelgarcia) [inline spikeinterface/curation/curation_tools.py:46]: And also this.

**Comment by alejoe91** [comment]
@samuelgarcia can you apply the suggested changes?

**Comment by samuelgarcia** [comment]
@DradeAW I pushed something to avoid playing with the random state directly.
numpy offer an object API for this.

I reading your implementtion again. I think it will be ultra slow no ? you are remove spikes one by one.

Could we just random N number (0 or 1) and add theses to `duplicates` vector and then remove then all at once ?
It would random left of right no ?

the recurssive apporach here have a very high cost.

File: spikeinterface/curation/curation_tools.py (hunk 3 above)
Comment (samuelgarcia) [reply spikeinterface/curation/curation_tools.py]: Done

File: spikeinterface/curation/curation_tools.py (hunk 2 above)
Comment (samuelgarcia) [reply spikeinterface/curation/curation_tools.py]: Done

File: spikeinterface/curation/curation_tools.py (hunk 1 above)
Comment (samuelgarcia) [reply spikeinterface/curation/curation_tools.py]: done

**Comment by DradeAW** [comment]
The left or right is what was done previously. It works on the first pass but not afterwards, otherwise you can stumble into the case where all spikes are removed.
I agree that the current implementation isn't very pretty, but generating a random number is actually quite fast.

If you have a better implementation in mind that works I'm all for it!

**Comment by DradeAW** [comment]
I'm ok with it!

**Comment by alejoe91** [comment]
ok, merging!

**Review by alejoe91 — APPROVED** [review]

**Pull request merged**